pipeline/translate/merge-corpus.sh

```diff
@@ -21,9 +23,9 @@ mkdir -p "${tmp_dir}"
 cat <(pigz -dc "${src1}") <(pigz -dc "${src2}") | pigz >"${tmp_dir}/original.src.gz"
 cat <(pigz -dc "${trg1}") <(pigz -dc "${trg2}") | pigz >"${tmp_dir}/original.trg.gz"
 
-echo "#### Shuffling"
+echo "#### Deduplicating"
 paste <(pigz -dc "${tmp_dir}/original.src.gz") <(pigz -dc "${tmp_dir}/original.trg.gz") |
-  shuf |
```

Comment:
I would still leave shuffling because sometimes marian is out of memory on big datasets and I have to shuffle batches. See https://github.com/mozilla/firefox-translations-training/issues/21. Just removing `--shuffle-in-ram` doesn't help.

Reply:
marian ALWAYS shuffles its input, even if the corpora is not in ram. Shuffling is done before the start of the training and then at every epoch. The only way to prevent shuffling is to use the `shuffle: none` option or `no-shuffle: true`:
https://github.com/marian-nmt/marian-dev/blob/a365bb5ce99135eab29ffe378e0c6c9fb9bf0c1b/src/common/config_parser.cpp#L426

External shuffle is completely redundant.

Reply:
```
--shuffle TEXT=data                   How to shuffle input data (data: shuffles data and sorted 
                                      batches; batches: data is read in order into batches, but 
                                      batches are shuffled; none: no shuffling).
                                      ```
so if I have to use `--shuffle batches` the data will be read in order and shuffled inside the batch if I get it right. wouldn't training quality benefit from preshuffled data in this case?

Reply:
again, I doesn't matter because the batches would be shuffled anyways. The important bit for training is to randomise the sequence in which the data is processed. Even if you use `--shuffle batches` the data is still shuffled, because input sentences are sorted by order before being fed into mini-batches. At the moment, the default behaviour `data` that is used makes shuffling redundant.

I am not sure what use case `shuffle: batches` covers, maybe for multilingual model training where  the order in which the data is seen matters a lot more.

Reply:
It covers issue #21. I have to use it on 128 GB GPUs for Italian and Portuguese for example. So, the data is shuffled only inside the batch, right? But data is read in order and the sequence of batches is not shuffled. It means it will be trained on corpus1, corpus2, corpus3 etc sequentially which is bad because the data inside every batch will be less diverse and it's easier to stuck in local minima. Do I misunderstand how it works?

Reply:
The data is stored externally in the `tmp` directory (`-T tmp`) and shuffled there in case it can't be in ram. `--shuffle-in-ram` does not modify the behaviour, it only speeds up the shuffling if you have enough RAM.
The way training works it:
1) Shuffle all input data You can see how this happens here: https://github.com/marian-nmt/marian-dev/blob/master/src/data/corpus.cpp#L205
2) Construct batches from the shuffled input, by reading `--maxi-batch` sentences and sorting them by length.
3) Training
The behaviour which you are describing is similar (but not exactly) to `shuffle: batches`, but the sequences in batches is always shuffled. Does this answer the question?

Reply:
Ok, I got it about the shuffle in ram option. I'm trying to understand how batching works. I'm looking at https://github.com/marian-nmt/marian-dev/blob/master/src/data/batch_generator.h

So here it says not to shuffle the corpus if it's not in `data` mode.
https://github.com/marian-nmt/marian-dev/blob/master/src/data/batch_generator.h#L315 

Then it reads data in order, forming maxi batches. Sentences are ordered inside a maxi batch in a priority queue. Then mini batches are formed from a maxi batch. We read data into maxi batch until we reach `maxi-batch=1000`. The size of a mini-batch is dynamic based on `mini-batch-fit: True` but not bigger than `mini-batch: 1000`. 

Then, when maxi-batch is full we shuffle mini-batches here https://github.com/marian-nmt/marian-dev/blob/aafe8fb5ca8f613f52da7589a72e1a647d51f820/src/data/batch_generator.h#L231.

Does all this mean that maxi batches (1M sentences max) are still read in order and not shuffled or I'm missing something?

Reply:
I'm not familiar with `batch` mode, I have not used it.
> The size of a mini-batch is dynamic based on mini-batch-fit: True but not bigger than mini-batch: 1000.

Not entirely correct, as there is a secondary logic in the training scheduling that may decrease or increase the effective mini-batch size (this is controlled by the `optimiser-delay: n` parameter.)

> Does all this mean that maxi batches (1M sentences max) are still read in order and not shuffled or I'm missing something? 

As far as I understand (and I may be wrong here), if running in `batch` mode (which we aren't), ALL batches will be shuffled at the start of each epoch, not just the first 1M, but they will contain the same data. They should be saved to temporary files and read in in different order at the start of every training. The 1M sentence limit is the RAM usage limit (eg the amount you are currently holding in memory/processing). Shuffling is not limited by it.

We should be using `data` mode of batching (and we are).

Reply:
Well, I have to use batch mode sometimes as a workaround as I pointed out in #21. Nothing else helped me with OOM  on student training. The pipeline has to be scalable to dataset size and I even considered switching it to batch mode everywhere to prevent this situation on huge corpora for high-resource languages. 

So unless we are 100% sure that removing the pre-shuffling won't affect model quality we should leave it as is. This operation is harmless and doesn't take that much time or resources compared to training or decoding, so I consider this removing a micro-optimization that doesn't help much anyway.

Reply:
Do you have this particular setup somewhere? Could I try to reproduce it? Marian shouldn't ever consume that much memory if it is not shuffling in ram. If it does it is a bug that we need to report and fix it. The whole reason files are written to a temporary directory is precisely for these scenarios, where memory is limited.

As for the data shuffling, without shuffling batches between epochs, shuffling does very little for translation quality.  The really important bit is to shuffle between epochs, and thinking about it, if the model sees always the same mini-batches, but in different order, that is also suboptimal for the training. This is especially true for student model training which should go for multiple epochs.

My problem with `shuf` is that it makes the pipeline non-deterministic meaning two runs on the same machine will result in slightly different models. I guess it's not a problem in production where you don't care for deterministic results. If you really want shuf there,  I'll put it back, but I am concerned that #21 happens at all.

Reply:
> Marian shouldn't ever consume that much memory if it is not shuffling in ram

I agree, this looks like a marian bug. I can try to reproduce it again when I reach student training. If it doesn't reproduce, we can remove shuffling. If it does, we can report a marian bug but leave this as a workaround until it's fixed. 

The setup is to train a student model with our current settings on 330M+ sentences and 128 GB GPU. It is pretty easy to get for high-resource languages + back translations + forward translation of 10 mono news-crawl datasets.

> My problem with shuf is that it makes the pipeline non-deterministic

I see, determinist is a different issue and it can be solved with `shuf --random-source=some_file`.

Reply:
I have some semblance of deterministic shuf in, let's see if it works